[extension/basicauth] Implement `configauth.ClientAuthenticator`

neelayu opened the pull request:
**Description:**
Enhancement: `basicauth` extension now supports client auth.

**Testing:** Manual and Unit Tests added.

neelayu committed:
Add support for client basic auth

neelayu committed:
Change Readme

neelayu committed:
Add changelog entry

jpkrohling commented:
cc @gouthamve

jpkrohling commented on extension/basicauthextension/README.md:
```diff
@@ -27,16 +27,17 @@ receivers:
 exporters:
-  logging:
-    logLevel: debug
+  otlphttp:
```
How about `otlp`? It should be used whenever possible instead of `otlphttp`, from what I understand.

neelayu replied:
currently the intention is to support only http client

jpkrohling commented on extension/basicauthextension/extension.go:
```diff
@@ -38,50 +39,54 @@ var (
 		return fmt.Errorf("read htpasswd content: %w", err)
 	}
-
+	ba.userPassPair = buff.String()
```
Am I right to assume that this is getting the username and password from the htpasswd file? If so, that's not right.

jpkrohling commented on extension/basicauthextension/extension.go:
```diff
@@ -177,3 +193,29 @@ func (a *authData) GetAttribute(name string) interface{} {
+}
+
+func (ba *basicAuth) PerRPCCredentials() (creds.PerRPCCredentials, error) {
+	return nil, nil
```
Why are you not implementing this?

neelayu replied:
Is this for gRPC?

jpkrohling replied:
Yes

jpkrohling commented on extension/basicauthextension/extension_test.go:
```diff
@@ -193,3 +206,108 @@ func TestBasicAuth_SupportedHeaders(t *testing.T) {
+			Inline: creds,
+		},
+	})
+	assert.NotNil(t, ext)
```
This should be a `require`: everything that is required for setting up the stage of the test should use `require`. When you are exercising the test subject, use `assert`.

neelayu replied:
Thanks for the tip. Reworked.

jpkrohling commented on extension/basicauthextension/extension.go:
```diff
@@ -38,50 +39,54 @@ var (
 }
 
-func newExtension(cfg *Config) (configauth.ServerAuthenticator, error) {
+func newExtension(cfg *Config) (*basicAuth, error) {
```
I think I would prefer the factory to determine whether it needs a client authenticator or a server authenticator (see `createExtension` in the factory.go).

I also think this extension has two different sets of configuration options:

- username and password for the client authenticator, with only one subject
- htpasswd for the server authenticator, with potentially multiple subjects

So, if both the username and password are set, this takes the shape of a client authenticator, otherwise it's a server authenticator. The extension cannot be both at the same time. As an operator, I would find it confusing to have one extension instance with the two facets at the same time.

neelayu replied:
Good points. I have incorporated that logic. I have also enforced that this extention will only act as either a client auth or server auth. If a user needs both, there should be two. Kindly review.

neelayu committed:
separate out client and server authenticator

svrakitin commented on extension/basicauthextension/README.md:
```diff
@@ -1,47 +1,58 @@
+
+- `htpasswd.file`:  The path to the htpasswd file.
+- `htpasswd.inline`: The htpasswd file inline content.
+- `htpasswd.client_auth`: Single username password combination in the form of `username:password` for client authentication.
+
+To configure the extension as a server authenticator, either one of `htpasswd.file` or `htpasswd.inline` has to be set. If both are configured, `htpasswd.inline` credentials take precedence.
+
+To configure the extension as a client authenticator, `htpasswd.client_auth` has to be set.
```
Shouldn't it be like this according to the actual code?

```suggestion
- `client_auth`: Single username password combination in the form of `username:password` for client authentication.

To configure the extension as a server authenticator, either one of `htpasswd.file` or `htpasswd.inline` has to be set. If both are configured, `htpasswd.inline` credentials take precedence.

To configure the extension as a client authenticator, `client_auth` has to be set.
```

neelayu replied:
Oh yes. Nice catch. Thanks for reviewing 😁

neelayu committed:
Fix Readme documentation

Co-authored-by: Stepan Rakitin <stepanr@mailbox.org>

neelayu committed:
Merge branch 'main' into client-basicauth

mar4uk commented on extension/basicauthextension/README.md:
```diff
@@ -1,47 +1,58 @@
+  
+  basicauth/client:
+    htpasswd:
+      client_auth: username:password
```
shouldn't it be:
```
basicauth/client:
   client_auth: username:password
```
?

neelayu committed:
Update Readme.md

neelayu committed:
modify internal components test file

neelayu commented:
@jpkrohling can you give it a one more glance at this. Thanks

jpkrohling commented on extension/basicauthextension/README.md:
```diff
@@ -1,47 +1,57 @@
         ${BASIC_AUTH_USERNAME}:${BASIC_AUTH_PASSWORD}
+  
+  basicauth/client:
+    client_auth: username:password
```
This is not correct, though. According to the actual code, it should be something like:

```
client_auth:
  username: jdoe
  password: secret
```

neelayu replied:
My bad. Actually I changed the logic after your comments in the first draft and I missed to correct the Readme at many places. Thank you.

jpkrohling commented on extension/basicauthextension/extension.go:
```diff
@@ -20,39 +20,66 @@ import (
+	), nil
+}
+
+func (ba *basicAuth) clientStart(_ context.Context, _ component.Host) error {
```
If you don't need it, don't implement it.

neelayu replied:
Noted

jpkrohling commented on extension/basicauthextension/extension.go:
```diff
@@ -81,6 +108,10 @@ func (ba *basicAuth) start(ctx context.Context, host component.Host) error {
 	return nil
 }
 
+func (ba *basicAuth) shutdown(ctx context.Context) error {
```
Same here: if there's nothing to shut down, you don't need to implement this.

neelayu replied:
Noted.

jpkrohling commented on extension/basicauthextension/extension_test.go:
```diff
@@ -193,3 +206,81 @@ func TestBasicAuth_SupportedHeaders(t *testing.T) {
+	require.NotNil(t, ext)
+	require.NoError(t, err)
+
+	require.Nil(t, ext.Start(context.Background(), componenttest.NewNopHost()))
```
```suggestion
	require.NoError(t, ext.Start(context.Background(), componenttest.NewNopHost()))
```

jpkrohling commented on extension/basicauthextension/extension_test.go:
```diff
@@ -193,3 +206,81 @@ func TestBasicAuth_SupportedHeaders(t *testing.T) {
+	resp, err := c.RoundTrip(&http.Request{Header: orgHeaders})
+	assert.NoError(t, err)
+	assert.Equal(t, expectedHeaders, resp.Header)
+	assert.Nil(t, ext.Shutdown(context.Background()))
```
```suggestion
	assert.NoError(t, ext.Shutdown(context.Background()))
```

jpkrohling replied:
This pattern repeats itself in other parts of this PR.

jpkrohling commented on extension/basicauthextension/extension_test.go:
```diff
@@ -193,3 +206,81 @@ func TestBasicAuth_SupportedHeaders(t *testing.T) {
+	assert.Nil(t, ext.Shutdown(context.Background()))
+}
+
+func TestBasicAuth_ClientInValid(t *testing.T) {
```
```suggestion
func TestBasicAuth_ClientInvalid(t *testing.T) {
```

jpkrohling commented on extension/basicauthextension/extension.go:
```diff
@@ -177,3 +208,28 @@ func (a *authData) GetAttribute(name string) interface{} {
+}
+
+func (ba *basicAuth) RoundTripper(base http.RoundTripper) (http.RoundTripper, error) {
+	if strings.Contains(ba.clientAuth.Username, ":") {
```
Is that the only thing that a username should not contain?

neelayu replied:
I was referring to this-
https://en.wikipedia.org/wiki/Basic_access_authentication (See client side)

jpkrohling replied:
I understand that, but is that the only char you have to guard against?

neelayu replied:
Based on [this](https://tools.ietf.org/html/rfc2617#page-6) RFC2617 Specification, HTTP Basic Authentication username can contain any TEXT excluding the symbol `:`

jpkrohling replied:
Thanks for the confirmation!

neelayu committed:
Apply suggestions from code review

Co-authored-by: Juraci Paixão Kröhling <juraci.github@kroehling.de>

neelayu committed:
add rpc support

neelayu committed:
Merge branch 'client-basicauth' of https://github.com/neelayu/opentelemetry-collector-contrib into client-basicauth

neelayu committed:
review comments and minor fixes

neelayu commented:
> This should then be clearly stated in the readme then. I'm sure I'm missing something obvious, but why can't this be used with gRPC? It would be similar to the bearer token auth, like this:

You are correct. It is indeed similar to bearer auth. The only reason I did not implement it was because it was written HTTP Basic Auth in the readme before. But sure it works for gRPC as well. I have removed HTTP from the readme now and also implemented it for gRPC. Thank you for reviewing it.

jpkrohling commented on extension/basicauthextension/config.go:
```diff
@@ -14,7 +14,16 @@
+
+var (
+	errNoCredentialSource    = errors.New("no credential source provided")
+	errMultipleAuthenticator = errors.New("only one of `htpasswd` or `client_auth` can be specified")
```
nit: `errMultipleAuthenticators` (plural)

neelayu replied:
Done

jpkrohling commented on extension/basicauthextension/extension.go:
```diff
@@ -20,39 +20,60 @@ import (
+		clientAuth: cfg.ClientAuth,
+	}
+	return configauth.NewClientAuthenticator(
+		configauth.WithClientRoundTripper(ba.RoundTripper),
```
You don't need the `ba.RoundTripper` nor the `ba.PerRPCCredentials` to be public.

neelayu replied:
Thanks for the tip. Done

jpkrohling commented on extension/basicauthextension/extension.go:
```diff
@@ -177,3 +198,50 @@ func (a *authData) GetAttribute(name string) interface{} {
 	return []string{"username", "raw"}
 }
+
+type perRPCAuth struct {
```
A comment here would be nice to have.

neelayu replied:
Done

jpkrohling commented on extension/basicauthextension/extension.go:
```diff
@@ -177,3 +198,50 @@ func (a *authData) GetAttribute(name string) interface{} {
+}
+
+func (ba *basicAuth) PerRPCCredentials() (creds.PerRPCCredentials, error) {
+	if strings.Contains(ba.clientAuth.Username, ":") {
```
You probably don't need this check here, do you?

neelayu replied:
Are you suggesting this check shouldn't be here and be somewhere else or not present at all?

jpkrohling replied:
The constraint is only about HTTP headers, not sure whether the gRPC header would have the same limitation. Leave it here for the moment, I think it makes more sense to do it here than not have it checked at all.

jpkrohling reviewed (commented):
This is looking good, just a couple of minor comments.

neelayu committed:
address comments and make func signatures private

neelayu committed:
gofmt lint error fix
